### PR TITLE
feat(smart-ngrx): reduce the number of repaint and re retrievals

### PR DESCRIPTION
**Pull request by DaveMBush**
# Issue Number: #591 

# Body

Made refinements to the code to reduce the number of repaints and calls to the server.

## Summary by CodeRabbit

- **New Features**
	- Improved performance of the tree component by reducing unnecessary updates through a debounce mechanism.
	- Enhanced reliability of data processing in the tree component service with additional validation checks.
	- Optimized loading process by filtering out unnecessary actions in the action service.
	- Added new test coverage for the `loadByIds()` method to ensure correct behavior under various conditions.
	- Introduced checks in the `VirtualArray` class to prevent modifications to frozen arrays.

- **Bug Fixes**
	- Prevented potential errors by ensuring the tree component service exits early when no valid data is available.

- **Refactor**
	- Simplified effects management by removing outdated loading effects, enhancing maintainability.
	- Enhanced clarity in the `load-by-indexes-effect` function by removing unnecessary comments.

- **Tests**
	- Improved test structure and coverage for the `VirtualArray` class, ensuring edge cases are validated.

### CODE REVIEW

File: libs/smart-ngrx/src/selector/virtual-array.class.ts [hunk 1]
```diff
@@ -46,6 +46,10 @@ export class VirtualArray<
               childField,
             }),
           );
+          if (Object.isFrozen(this.rawArray)) {
+            this.rawArray = [...this.rawArray];
+          }
+          this.rawArray[+prop] = `indexNoOp-${prop}`;
```
[the remark this answers is not in the document]
Comment (DaveMBush) [reply libs/smart-ngrx/src/selector/virtual-array.class.ts:52]: rawArray is a string array. spread is appropriate here

File: apps/demo/src/app/shared/components/tree/tree-component.service.ts [hunk 2]
```diff
@@ -31,7 +31,12 @@ export class TreeComponentService {
   applyRange(): void {
     const component = this.component;
     assert(!!component, 'component is null');
-    if (component.location() === undefined) {
+    // no use in painting if there is nothing to paint
+    if (
+      component.location() === undefined ||
+      component.location() === null ||
+      component.location()!.departments.length === 0
+    ) {
```
[the remark this answers is not in the document]
Comment (DaveMBush) [reply apps/demo/src/app/shared/components/tree/tree-component.service.ts:39]: this happens to already be covered by existing tests

File: libs/smart-ngrx/src/actions/action.service.ts [hunk 3]
```diff
@@ -259,11 +259,19 @@ export class ActionService {
...
+      );
+      if (ids.length === 0) {
+        return;
+      }
+      this.store.dispatch(
+        this.actions.loadByIds({
+          ids,
+        }),
+      );
+    });
```
[the remark this answers is not in the document]
Comment (DaveMBush) [reply libs/smart-ngrx/src/actions/action.service.ts:274]: I'm going to refactor the code to not need tests in the next commit to this PR

File: libs/smart-ngrx/src/effects/effects-factory/load-by-ids-effect.function.ts [hunk 4]
```diff
@@ -27,17 +32,18 @@ export function loadByIdsEffect<T extends SmartNgRXRowBase>(
     /* istanbul ignore next -- default value, not really a condition */
     actions$ = inject(Actions),
     /* istanbul ignore next -- default value, not really a condition */
-    actionService = inject(effectServiceToken),
+    effectService = inject(effectServiceToken),
```
[the remark this answers is not in the document]
Comment (DaveMBush) [reply libs/smart-ngrx/src/effects/effects-factory/load-by-ids-effect.function.ts:35]: I've checked the places you've mentioned and they are all appropriately named now.
